plugins/qcheck-stm/doc/index.mld

```diff
@@ -0,0 +1,275 @@
+https://ocaml-multicore.github.io/multicoretests/ }QCheck-STM} to perform
+model-based state-machine testing of a module by using the model of its {{:
+https://ocaml-gospel.github.io/gospel/ }Gospel} specifications.
+
+In order to be able to generate the STM module, the plugin will need some information:
+
+{ol {- What type do we want to test? This is called {e system under test} or {e
+       sut} by QCheck-STM}
+    {- How to generate a value of this type?}
+    {- What is the model of this type? This is what is taken as the {e state} by QCheck-STM}
+    {- How to generate the said model?}
+    {- How does the model change when calling a function?} }
+
+Answering these five questions is done part in the command-line arguments and
+part in the gospel specifications that you will have to write in a specific
+way.
+
+This tutorial aims at showing how to write gospel specifications to your
+modules in order to test them using [QCheck-STM], and then how to call the
+[ortac] command-line tool with the [qcheck-stm] plugin.
+
+{1 How to write gospel specifications?}
+
+In order to use the [qcheck-stm] plugin, the module you want to test must
+contains the declaration of the type you want to use QCheck-STM on. Here is an
+example with a simple fixed-size container type. You'll notice the Gospel
+specifications under the type declaration, in the special Gospel comments.
+These specifications give two models to the type, one non mutable and one
+mutable. These models are necessary (or at least some models) in order to give
+enough information to the [qcheck-stm] to build the functional state that
+QCheck-STM will be using.
+
+{@ocaml file=example.mli,part=type-decl[
+type 'a t
+(*@ model size : int
+    mutable model contents : 'a list *)
+]}
+
+The second two pieces of information you need to give are about building both a
+value of the type we want to test and the corresponding model. This is the role
+of the [make] function here along with its specification. The specification
+will have to give a value to each of the models that were given to the type ['a
+t] in its specifications. Here, this means we need to give a value to [t.size]
+and to [t.contents]. The [checks] clause is part of the specification of the
+function, but it won't be used by the plugin to generate any code. You can give
+more information that the plugin need, for example if you are also using
+another tool based on Gospel specifications.
+
+{@ocaml file=example.mli,part=make[
+val make : int -> 'a -> 'a t
+(*@ t = make i a
+    checks i >= 0
+    ensures t.size = i
+    ensures t.contents = List.init i (fun j -> a) *)
+]}
+
+Now that [qcheck-stm] is able to generate an initial value for the type under
+test and its model, we can turn our attention to the functions we will want to
+test. Here is the example of the classic [set] function. You can see again the
+[checks] clause. Here again, it will be ignored by the tool. Then there is the
```

Comment:
In that case, the `checks` clause will be tested to check the function did raise an exception.